typeid/typeid-go/typeid.go

```diff
@@ -10,44 +10,49 @@ import (
 )
 
 // TypeID is a unique identifier with a given type as defined by the TypeID spec
-type TypeID struct {
-	prefix string
+type TypeID[P Prefix] struct {
+	prefix P
 	suffix string
 }
 
-// Nil represents an the null TypeID
-var Nil = TypeID{
-	prefix: "",
-	suffix: "00000000000000000000000000",
+// Nil represents a the null TypeID
+func Nil[P Prefix]() TypeID[P] {
+	// TODO: See if we can make Nil == default initialization of a TypeID struct
+	var prefix P
+	return TypeID[P]{
+		prefix: prefix,
+		suffix: "00000000000000000000000000",
+	}
 }
 
-// New returns a new TypeID with the given prefix and a random suffix.
-// If you want to create an id without a prefix, pass an empty string.
-func New(prefix string) (TypeID, error) {
-	return From(prefix, "")
+// New returns a new TypeID of the given type.
+// The prefix is inferred from the type, and the suffix is random.
+func New[P Prefix]() (TypeID[P], error) {
```

Comment:
If I do `New[AnyPrefix]()` this will return an id of the form `_123` right? That seems wrong (i.e. it would be missing the prefix completely)

Reply:
No, it would return `123` with no underscore in front (unless you're saying I have a bug in handling the empty prefix?). We do allow for an empty prefix in the spec and in all other implementations in different languages.